packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,479 @@
+        // prove. We can only monitor the 'usedCapacity' in case we increase it once currentCapacity 
+        // deducted, but that increases writes. So simply we shall allow 'exit' with currentCapacity
+        // Then when (if) they reenter, they need to specify their actually available capacity in the exact moment
+        // they are reentering. If possible, try to avoid extra complexity into the contracts.
+        //uint16 usedCapacity;
+    }
+
+    // TOP PROVERS LIST
+    TopProver[32] public topProvers; // 32/4 = 8 slots
+    // EXITING PROVER MAPPING
+    mapping(address prover => ExitingProver) exitingProvers;
+
+    // Id mappings
+    mapping(uint8 id => address) idToProver;
+    mapping(address prover => uint8 id) proverToId;
+
+    // Exit period. Might fine tune later on !
+    uint256 public constant EXIT_PERIOD = 1 weeks;
+
+    // 500 means 5% if 10_000 is 100 %
+    uint256 public constant SLASH_AMOUNT_IN_BP = 500; 
+
+    uint256 public WEI_TO_1_TKO = 10e8;
+
+    uint256[100] private __gap;
+
+    event ProverEntered(
+        address prover, uint32 amount, uint16 rewardPerGas, uint16 capacity
+    );
+
+    event ProverChangedParameters(address prover, uint32 newBalance, uint16 newReward, uint16 newCapacity);
+
+    event ProverKickeOutByWithAmount(address kickedOut, address newProver, uint32 totalAmount);
+
+    event ProverExitRequested(address prover, uint64 timestamp);
+
+    event ProverExited(address prover, uint64 timestamp);
+
+    event ProverSlashed(address prover, uint32 newBalance);
+
+    modifier onlyProtocol() {
+        require(
+            AddressResolver(this).resolve("taiko", false) != msg.sender,
+            "Only Taiko L1 can call this function"
+        );
+        _;
+    }
+
+    modifier onlyActiveProver(address prover) {
+        // Either within top provers (rewardPerGas != 0 serves as a check), or still exiting
+        require(
+            topProvers[proverToId[prover]].rewardPerGas != 0
+            || 
+            ( exitingProvers[prover].exitRequestedAt != 0
+            && exitingProvers[prover].exitRequestedAt + EXIT_PERIOD < block.timestamp),
+            "Only active provers can call this function"
+        );
+        _;
+    }
+
+    /**
+     * Initialize the rollup.
+     *
+     * @param _addressManager The AddressManager address.
+     */
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    // A demo how to optimize the assignProver by using only 8 slots. It's still
+    // a lot of slots tough.
+    function assignProver(
+        uint64 blockId,
+        uint32 feePerGas
+    )
+        external
+        returns (address prover, uint32 rewardPerGas)
+    {
+        // readjust each prover's rate
+        uint256[32] memory weights;
+        uint256 totalWeight;
+
+        // Iterate over in-memory array
+        TopProver[32] memory mTopProvers = topProvers;
```

Comment:
We should load topProvers one by one, not all at once, assuming there is only one prover in the list...